### ISSUE TITLE
Connection Issue with LakeShore Model 335 Temperature Controller

### ISSUE THREAD

**Issue by juggernaut545** [lakeshorecryotronics/python-driver#6]
Hello. I am attempting to connect my Model 335 temperature controller to the computer. I have connected the USB and the connection is read by PyVisa. I run into errors, however, when I attempt to run code with it. I have pip installed Lakeshore and also installed the corresponding driver from the Lakeshore website. 

Here is the code that I am attempting to run (I am just testing out some methods to get a feel for it). 
```

import pyvisa
from lakeshore import Model335
from lakeshore.model_335 import *

rm = pyvisa.ResourceManager()
print(rm.list_resources())

my_instrument = Model335(57600)

my_instrument.set_display_setup(Model335DisplaySetup.TWO_INPUT_A)
```
I used this example code from the [readthedocs website](https://lake-shore-python-driver.readthedocs.io/en/latest/model_335.html).
Here is the error output:
```
('ASRL3::INSTR', 'GPIB0::8::INSTR')
Traceback (most recent call last):

  File ~\Documents\temperaturetestcode.py:15 in <module>
    my_instrument = Model335(57600)

  File ~\anaconda3\lib\site-packages\lakeshore\model_335.py:329 in __init__
    TemperatureController.__init__(self, serial_number, com_port, baud_rate, timeout, ip_address,

  File ~\anaconda3\lib\site-packages\lakeshore\temperature_controllers.py:281 in __init__
    GenericInstrument.__init__(self, serial_number, com_port, baud_rate, 7, 1, serial.PARITY_ODD,

  File ~\anaconda3\lib\site-packages\lakeshore\generic_instrument.py:77 in __init__
    self.connect_usb(serial_number, com_port, baud_rate, data_bits, stop_bits, parity,

  File ~\anaconda3\lib\site-packages\lakeshore\generic_instrument.py:189 in connect_usb
    self.device_serial = serial.Serial(port.device,

  File ~\anaconda3\lib\site-packages\serial\serialwin32.py:33 in __init__
    super(Serial, self).__init__(*args, **kwargs)

  File ~\anaconda3\lib\site-packages\serial\serialutil.py:244 in __init__
    self.open()

  File ~\anaconda3\lib\site-packages\serial\serialwin32.py:64 in open
    raise SerialException("could not open port {!r}: {!r}".format(self.portstr, ctypes.WinError()))

SerialException: could not open port 'COM3': PermissionError(13, 'Access is denied.', None, 5)
```

I am not sure as to what is causing this issue. I was wondering if it might be something to do with where the corresponding files/drivers were installed, but the error line ```SerialException: could not open port 'COM3': PermissionError(13, 'Access is denied.', None, 5)``` does not seem to be related to that, so I am unsure. Please let me know if you have any suggestions. I appreciate all and any help, and thank you for your time.

**Comment by Noah-Lake-Shore-Cryotronics** [comment]
This error most often occurs when another piece of software is already using the port. The connection can only be used by one piece of software at a time. Please close out of any other programs and/or disconnect and reconnect the Model 335 from your computer to force the connection to drop.

Please let me know if this fixes the problem.

**Comment by juggernaut545** [comment]
Thank you for the advice. 
I reinstalled the drivers and it seemed to have fixed the error that the COM port could not be opened, but now I am encountering a new error.

Here is a screenshot of the only COM port that is in use according to device manager. 
![image](https://user-images.githubusercontent.com/29962836/201445497-9aa406be-2421-4efa-b4e6-471090ce36fa.png)

This temperature controller is what I am trying to connect to using Pyvisa. I also have a lock-in amplifier connected. Here is the basic code that I am running: 
```import pyvisa
from lakeshore import Model335
from lakeshore.model_335 import *

rm = pyvisa.ResourceManager()
print(rm.list_resources())

m1 = rm.open_resource('GPIB0::8::INSTR')
m2 = rm.open_resource('ASRL3::INSTR')
```

Here is my output again:
```('ASRL3::INSTR', 'GPIB0::8::INSTR')
Traceback (most recent call last):

  File ~\Documents\temperaturetestcode.py:16 in <module>
    m2 = rm.open_resource('ASRL3::INSTR')

  File ~\anaconda3\lib\site-packages\pyvisa\highlevel.py:3284 in open_resource
    res.open(access_mode, open_timeout)

  File ~\anaconda3\lib\site-packages\pyvisa\resources\resource.py:278 in open
    self.session, status = self._resource_manager.open_bare_resource(

  File ~\anaconda3\lib\site-packages\pyvisa\highlevel.py:3209 in open_bare_resource
    return self.visalib.open(self.session, resource_name, access_mode, open_timeout)

  File ~\anaconda3\lib\site-packages\pyvisa\ctwrapper\functions.py:1851 in open
    ret = library.viOpen(

  File ~\anaconda3\lib\site-packages\pyvisa\ctwrapper\highlevel.py:226 in _return_handler
    return self.handle_return_value(session, ret_value)  # type: ignore

  File ~\anaconda3\lib\site-packages\pyvisa\highlevel.py:251 in handle_return_value
    raise errors.VisaIOError(rv)

VisaIOError: VI_ERROR_RSRC_BUSY (-1073807246): The resource is valid, but VISA cannot currently access it.
```

I have tried looking online but the only answers I could find were those that said to make sure nothing else was connected to the serial port. In device manager, it does not show anything connected to COM port 3. I am at a loss as to how to proceed. Please let me know if you need any further information.

Thank you in advance.

**Comment by juggernaut545** [comment]
It seems as the code works on the first time I run it, but I encounter the ```SerialException: could not open port 'COM3': PermissionError(13, 'Access is denied.', None, 5)``` error comes whenever I run it again. Is there a way to circumvent this?

**Comment by Noah-Lake-Shore-Cryotronics** [comment]
I still think that the most likely explanation for these errors is a program running the background that is hogging the connection. Are you using any serial communications debugging tools to manually send commands to the instrument for example?

If you are using pyvisa instead of the driver to connect to the instrument, please ensure that your serial communication settings such as data bits, stop bits, and bit parity are configured correctly. One advantage of using the Lake Shore driver is that these settings are already baked in.

I am surprised by the issue you are having where the script throws an error whenever it is rerun. The connection to the com port should not remain open when the script is terminated. To help with debugging please let me know the following:

1. What development environment are you using to run your python script?
2. After restarting your computer to ensure no other programs using the connection, please run the following script multiple times in a row and let me know the result:

```
from lakeshore.model_335 import *

my_model_335 = Model335(57600)
```

**Comment by juggernaut545** [comment]
Thank you for the advice.

I am am not using any other serial communications debugging tools to manually send commands to the instrument. I have my lock-in amplifier connected with a GPIB-USB cable along with the 335 temperature controller cable (and some other usb cables for my mouse and keyboard). 

I am using Spyder (Anaconda) version 3.9.12. 

I restarted my computer and ran the code you provided, but I ran into the same error. On the first run, it was fine. However, upon rerunning it, I was met with the same error shown below. 
![image](https://user-images.githubusercontent.com/29962836/201972520-5c02bf10-027f-4195-8a97-147fcf6e500a.png)

Below is a screenshot of the only COM port listed in device manager. 
![image](https://user-images.githubusercontent.com/29962836/201973213-8b4df816-2c1b-4417-b586-67288ad03b6c.png)
Here is more specific information about the ports if they are helpful: 
![image](https://user-images.githubusercontent.com/29962836/201973542-f7b507cd-5f50-4187-a7f3-4937cc191123.png)

Is it possible that applications such as TekVisa or Ni-Visa are responsible for this? I closed out TekVisa and tried the above, and I do not believe the NI software was open (there were extraneous NI processes in task manager but I do not believe the application itself was open). 
Please let me know your suggestions. I greatly appreciate your time and help.
Thank you.

**Comment by Noah-Lake-Shore-Cryotronics** [comment]
I am not familiar with the runcell function but I suspect that what is happening is that the script is not being terminated and so the first established connection is not being closed. It would be equivalent to writing code like so:

```
from lakeshore.model_335 import *

my_model_335 = Model335(57600)

from lakeshore.model_335 import *

my_model_335 = Model335(57600)
```

This code will fail with the "access denied" error on the last line because the connection to the instrument is already in use.

Please try the following and let me know the result:
1. Exit the python interpreter in your console
2. Type `python <path to script>/untitled0.py` in the console and press enter. Assuming the interpreter is available at that location this should run your script.
3. Rerun the script multiple times with that same console command, I believe you will no longer encounter the access denied error on reruns.

**Comment by juggernaut545** [comment]
Thank you very much for your help. I was able to figure out a (rather foolish) workaround. 

 It did seem like the runcell option was not closing the connection and processes after the script itself ended. To get around this, I simply added a line at the end of the script that set the variable with the connection to another value.

For example:

```
my_model_335 = Model335(57600)
*(rest of code)*
my_model_335 = 0
```

This may cause further problems in the future, but for now it seems like an acceptable and convenient alternative rather than restarting the kernel every time or running it from the terminal. 

Thank you once again.

**Comment by juggernaut545** [comment]
Thank you very much for the advice. 
I believe that the runcell command not properly closing the connection after it was run was the source of the error as you said. I was able to figure out a (rather foolish) workaround. At the end of my code, I changed to variable holding the connection to another value, which I believe resulted in the termination of the connection.

For example:

```
from lakeshore import Model335

from lakeshore.model_335 import *


my_model_335 = Model335(57600)

*(rest of code)*

my_model_335 = 0

```

This is more convenient and easier than restarting the kernel every time or running it from the command line, although it might cause issues in the future. However, for now, it is an acceptable alternative.

Thank you very much for your advice and help. I greatly appreciate it.
